jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java

```diff
@@ -811,6 +805,75 @@ public static String getCanonicalNameForWebAppTmpDir(WebAppContext context)
         return StringUtil.sanitizeFileSystemName(canonicalName.toString());
     }
 
+    protected static String getResourceBaseName(Resource resource)
+    {
+        // Use File System and File interface if present
+        try
+        {
+            File resourceFile = resource.getFile();
+            if (resourceFile != null)
+            {
+                return resourceFile.getName();
+            }
+        }
+        catch (IOException e)
+        {
+            e.printStackTrace();
+        }
+
+        // Use URI itself.
+        URI uri = resource.getURI();
+        if (uri == null)
+        {
+            throw new RuntimeException("Unable to produce URI from resource: " + resource);
+        }
+        return getUriLastPathSegment(uri);
+    }
+
+    protected static String getUriLastPathSegment(URI uri)
```

Comment:
This seems to be an awful lot of trouble to go to just to make the name of a tmp directory. This portion of the whole name of the tmp dir could actually be "tmpymctempdir". Don't forget, the full name of the tmp dir is: "jetty-"+host+"-"+port+"-"+resourceBase+"-_"+context+"-"+virtualhost+"-"+randomdigits+".dir".

The point of putting the resourceBase in there I think was to just be more meaningful to a user looking at it. If it's truly this problematic we could just drop that portion.

Reply:
If you want a sane `resourceBase` segment, then this is what you want.

If you don't care about `jar:file://` based URI's, then just use the `Resource.getFile()` for `JarFileResource` which will return the jar filename, not the last segment in the JAR (like it was before this PR).

Reply:
The reason that `Resource.toURI().getPath()` returns null is that a URI of `jar:file://path/to/test.jar!/internal/path` has no path.
It's an external URI of scheme `jar:` and an internal URI of `file:` of which only the internal URI has a path.

Reply:
If you want this to be simpler, I can eliminate the entire `.getUriLastPathSegment(uri)` and only rely on `Resource.getFile()`.
Which will work for FileSystem, Path, File, `file:/` URIs, and `jar:file:/` URIs.

We would need a fallback in case there is a custom Resource implementation, and/or one that doesn't implement `Resource.getFile()` (like our `UrlResource`)

Reply:
What about we use Resource.getName() instead? All the Resource types implement that I think.

Reply:
`Resource.getName()` returns the full name of the resource, if the resource supports it.

* FileResource.getName() -> File.getAbsolutePath()
* PathResource.getName() -> Path.toAbsolutePath().toString()
* UrlResource - URL.toExternalForm()
* JarResource - URL.toExternalForm()
* JarFileResource - URL.toExternalForm()
* BadResource - URL.toExternalForm()
* EmptyResource - null
* ResourceCollection - first Resource.getName() that isn't null

That's not useful for us.

Reply:
And for completeness ... Using `Resource.getFile()` returns the File object if the resource supports it.

* FileResource.getFile() -> java.io.File
* PathResource.getFile() -> java.io.File
* UrlResource.getFile() -> null
* JarResource.getFile() -> null (this could be fixed to support it)
* JarFileResource.getFile() -> null (this could be fixed to support it)
* BadResource.getFile() -> null
* EmptyResource.getFile() -> null
* ResourceCollection.getFile() -> first Resource.getFile() that isn't null

Reply:
@joakime so it looks to me like we call Resource.getFile() first, if non-null, call File.getName(). Else, call Resource.getName() and strip off the last name element.